### ISSUE TITLE
Exception while invoking method 'testSend' Error: getaddrinfo ENOTFOUND

### ISSUE THREAD

**Issue by dialogik** [sandstorm-io/sandstorm#635]
Hi, I'm just setting up my Sandstorm. Now I do have an "@" sign in my username. So the STMP URL would be something like

    smtp://username@mydomain.tld:password@smtp.domain.tld

This throws an error:

    Exception while invoking method 'testSend' Error: getaddrinfo ENOTFOUND
        at Object.Future.wait (/programs/server/node_modules/fibers/future.js:398:15)
        at smtpSend (packages/sandstorm-email/email.js:101:1)
        at Object.SandstormEmail.rawSend (packages/sandstorm-email/email.js:176:1)
        at Object.SandstormEmail.send (packages/sandstorm-email/email.js:161:1)
        at [object Object].Meteor.methods.testSend (app/shared/admin.js:850:22)
        at maybeAuditArgumentChecks (packages/ddp/livedata_server.js:1617:1)
        at packages/ddp/livedata_server.js:648:1
        at [object Object]._.extend.withValue (packages/meteor/dynamics_nodejs.js:56:1)
        at packages/ddp/livedata_server.js:647:1
        at [object Object]._.extend.withValue (packages/meteor/dynamics_nodejs.js:56:1)
        - - - - -
        at errnoException (dns.js:37:11)
        at Object.onanswer [as oncomplete] (dns.js:124:16)

So I tried to encode the "@" sign...

    smtp://username%40mydomain.tld:password@smtp.domain.tld

... which throws the same error.

So what's the right way to use an "@" sign in my username?

**Comment by dialogik** [comment]
It actually also fails with Gmail's SMTP

    smtp://username:password@smtp.gmail.com

or

    smtp://username:password@smtp.gmail.com:465

Same error message as above.

**Comment by dialogik** [comment]
... and this also fails for `@smtp-relay.gmail.com` (same error).

**Comment by jparyani** [comment]
It seems your Sandstorm server is having trouble resolving hostnames. Could you verify that those hosts resolve outside of Sandstorm (ie. with `dig smtp.gmail.com`). If it's working, then try and restart the Sandstorm server (`sudo sandstorm restart`). A common issue that people run into is their /etc/resolv.conf changing, and Sandstorm caching an old version from when it started up.

**Comment by kentonv** [comment]
Note that `smtp.gmail.com` won't work as your Sandstorm SMTP server, because it is not a relay. That is, `smpt.gmail.com` only allows you to send mail as if from your own gmail account, whereas Sandstorm needs to send email from arbitrary addresses @ your server's hostname.

This seems like a common misunderstanding. Perhaps we should make the code explicitly detect this.

**Comment by dialogik** [comment]
@jparyani I indeed do get a `status: REFUSED` when I dig `smtp.gmail.com`, `smtp-relay.gmail.com` or my hosting provider's SMTP server. How can I resolve that?

**Comment by dialogik** [comment]
@kentonv Yeah, I noticed that while doing some research. That's why I chose the relay server of gmail.

**Comment by kentonv** [comment]
@dialogik Check your `/etc/resolv.conf`; it needs a `nameserver` line to declare a DNS server for lookups. It may be that the server it is using has some sort of restrictions. You can try using Google's public DNS instead:

    nameserver 8.8.8.8

**Comment by dialogik** [comment]
@kentonv Thanks for the heads-up. Due to corporate policies I cannot modify the nameserver. So now I'm using the corporate mail server and get a different error

    Exception while invoking method 'testSend' URIError: URI malformed

My URI looks like this

    smtp://myusername:ThisIsMyPassword%WithAPercentSignInIt@mail.domain.tld:465

and due to the `malformed` error I tried encoding the percent like this

    smtp://myusername:ThisIsMyPassword%25WithAPercentSignInIt@mail.domain.tld:465

but the error remains. Why is it malformed? And how can I fix this?

**Comment by paulproteus** [comment]
Dd you click "Save" at the bottom? For what it's worth, I just tried that
literal string on my Sandstorm and saved and it said it was OK. I tried
both of these:

smtp://myusername:ThisIsMyPassword%WithAPercentSignInIt@mail.domain.tld:465


smtp://myusername:ThisIsMyPassword%25WithAPercentSignInIt@mail.domain.tld
:465


I wonder if the password contains other characters that need to be encoded.
If so, this suggests we should show a username+password+host+port question,
rather than this compressed question.
